src/v3/recordingComputeUtils.ts

```diff
@@ -72,14 +76,126 @@ const markedInteractive = <RelationSchemasT>(
+  RelationSchemasT,
+  VariantsT extends string,
+>(
+  spanDef:
+    | SpanMatchDefinition<SelectedRelationNameT, RelationSchemasT, VariantsT>
+    | 'operation-start'
+    | 'operation-end'
+    | 'interactive'
+    | SpanMatcherFn<
+        NoInfer<SelectedRelationNameT>,
+        RelationSchemasT,
+        VariantsT
+      >,
+):
+  | SpanMatcherFn<SelectedRelationNameT, RelationSchemasT, VariantsT>
+  | string
+  | undefined {
+  // Handle string types (special matchers)
+  if (typeof spanDef === 'string') {
+    return spanDef
+  }
+
+  // Handle function types
+  if (typeof spanDef === 'function') {
+    return spanDef
+  }
+
+  // Handle object types
+  if (typeof spanDef === 'object' && spanDef !== null) {
+    const matcher = fromDefinition<
+      SelectedRelationNameT,
+      RelationSchemasT,
+      VariantsT
+    >(spanDef)
+
+    // Transfer top-level matchingIndex property
+    if (
+      'matchingIndex' in spanDef &&
+      spanDef.matchingIndex !== undefined &&
+      typeof matcher === 'function'
+    ) {
+      matcher.matchingIndex = spanDef.matchingIndex
+    }
+
+    return matcher
+  }
+
+  return undefined
+}
+
+/**
+ * Helper function to find matching spans according to a matcher and matching index
+ */
+function findMatchingSpan<
+  SelectedRelationNameT extends keyof RelationSchemasT,
+  RelationSchemasT,
+  VariantsT extends string,
+>(
+  matcher:
+    | ((
+        spanAndAnnotation: SpanAndAnnotation<RelationSchemasT>,
+      ) => boolean | undefined)
+    | SpanMatcherFn<
+        SelectedRelationNameT & keyof RelationSchemasT,
+        RelationSchemasT,
+        VariantsT
+      >,
+  recordedItemsArray: SpanAndAnnotation<RelationSchemasT>[],
+  context: TraceContext<SelectedRelationNameT, RelationSchemasT, VariantsT>,
+): SpanAndAnnotation<RelationSchemasT> | undefined {
+  if (typeof matcher !== 'function') return undefined
+
+  // For positive or undefined indices - find with specified index offset
+  if (
+    !('matchingIndex' in matcher) ||
+    matcher.matchingIndex === undefined ||
+    matcher.matchingIndex >= 0
+  ) {
+    let matchingIndex = 0
+    for (const spanAndAnnotation of recordedItemsArray) {
+      if (matcher(spanAndAnnotation, context)) {
+        if (
+          !('matchingIndex' in matcher) ||
+          matcher.matchingIndex === undefined ||
+          matcher.matchingIndex === matchingIndex
+        ) {
+          return spanAndAnnotation
+        }
+        matchingIndex++
+      }
+    }
+    return undefined
+  }
+
+  // For negative indices - collect all and pick from the end
+  const matches: SpanAndAnnotation<RelationSchemasT>[] = []
+  for (const spanAndAnnotation of recordedItemsArray) {
+    if (matcher(spanAndAnnotation, context)) {
+      matches.push(spanAndAnnotation)
+    }
+  }
+
+  const actualIndex = matches.length + matcher.matchingIndex
+  return actualIndex >= 0 && actualIndex < matches.length
+    ? matches[actualIndex]
+    : undefined
```

Comment:
Iterating everything is might be bad for performance, we'll want to iterate and match from the end - only as many times as specified in the `matchingIndex` (so if -1, we only need 1 match, if -2, we need the 2nd match).

It might be good to do this in `createTraceRecording` and pass in both `recordedItemsArray` and `recordedItemsArrayReversed`, so we don't reverse every time we need to calculate a computed span. I'm also wondering if we should sort `recordedItemsArrayReversed` by the end time...? For that matter, should `recordedItemsArray` be sorted by their start time? Maybe a question for #frontend-tracing channel.